From: Steven Rostedt <srostedt@redhat.com>
Subject: handle IRQ_PENDING for simple irq handler

With the IO-APIC pcix hack (level=>edge masking), we can receive
interrupts while masked. But these interrupts might be missed.

Also, normal "simple" interrupts might be missed too on leaving of
thread handler.

Signed-off-by: Steven Rostedt <srostedt@redhat.com>

---
 kernel/irq/manage.c |    7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

Index: linux-2.6.24-rt1/kernel/irq/manage.c
===================================================================
--- linux-2.6.24-rt1.orig/kernel/irq/manage.c	2008-01-25 15:08:17.000000000 -0500
+++ linux-2.6.24-rt1/kernel/irq/manage.c	2008-01-25 15:08:32.000000000 -0500
@@ -652,14 +652,17 @@ static void thread_simple_irq(irq_desc_t
 	unsigned int irq = desc - irq_desc;
 	irqreturn_t action_ret;
 
-	if (action && !desc->depth) {
+	do {
+		if (!action || desc->depth)
+			break;
+		desc->status &= ~IRQ_PENDING;
 		spin_unlock(&desc->lock);
 		action_ret = handle_IRQ_event(irq, action);
 		cond_resched_hardirq_context();
 		spin_lock_irq(&desc->lock);
 		if (!noirqdebug)
 			note_interrupt(irq, desc, action_ret);
-	}
+	} while (desc->status & IRQ_PENDING);
 	desc->status &= ~IRQ_INPROGRESS;
 }
 
